python-package/lightgbm/dask.py

```diff
@@ -338,6 +494,99 @@ def _train(
         for i in range(n_parts):
             parts[i]['init_score'] = init_score_parts[i]
 
+    # evals_set will to be re-constructed into smaller lists of (X, y) tuples, where
+    # X and y are each delayed sub-lists of original eval dask Collections.
+    if eval_set:
+
+        # find maximum number of parts in an individual eval set so that we can
+        # pad eval sets when they come in different sizes.
+        n_largest_eval_parts = max([x[0].npartitions for x in eval_set])
+
+        eval_sets = defaultdict(list)
+        if eval_sample_weight:
+            eval_sample_weights = defaultdict(list)
+        if eval_group:
+            eval_groups = defaultdict(list)
+        if eval_init_score:
+            eval_init_scores = defaultdict(list)
+
+        for i, (X_eval, y_eval) in enumerate(eval_set):
+            n_this_eval_parts = X_eval.npartitions
+
+            # when individual eval set is equivalent to training data, skip recomputing parts.
+            if X_eval is data and y_eval is label:
+                for parts_idx in range(n_parts):
+                    eval_sets[parts_idx].append(_DatasetNames.TRAINSET)
+            else:
+                eval_x_parts = _split_to_parts(data=X_eval, is_matrix=True)
+                eval_y_parts = _split_to_parts(data=y_eval, is_matrix=False)
+                for j in range(n_largest_eval_parts):
+                    parts_idx = j % n_parts
+
+                    # add None-padding for individual eval_set member if it is smaller than the largest member.
+                    x_e = eval_x_parts[j] if j < n_this_eval_parts else None
+                    y_e = eval_y_parts[j] if j < n_this_eval_parts else None
+
+                    if j < n_parts:
+                        eval_sets[parts_idx].append(([x_e], [y_e]))
+                    else:
+                        eval_sets[parts_idx][-1][0].append(x_e)
+                        eval_sets[parts_idx][-1][1].append(y_e)
```

Comment:
Oh yeah, no your reasoning is correct. Yes, this is the "round robin" thing you mentioned above, and it's due to the modulo on line 524: `parts_idx = j % n_parts`. So in your example, `worker_0` could get `[(X_e0, X_e2, ..., X_e10), (y_e0, y_e2, ..., y_e10)]`, leaving `worker_1` with `[(X_e1, X_e3, ..., X_e11), (y_e1, y_e3, ..., y_e11)]`. Until now I guess I have been disregarding data shuffling almost entirely since there's no guarantee in any point of this dask module that certain workers will be dispatched any individual part of the training set. But you're right, I mean there's no reason to purposefully disorganize the order of the parts. Distributing the data across workers in a random way is not the same as `shuffling`! Huh, I never really fully stopped to consider the differences between randomly distributing contiguous chunks versus simply building an equal allocation of random chunks.

A more intuitive approach might be to calculate before the outer loop exactly how many eval_set parts should be assigned to each worker in as equal an allocation as possible. Like in your example, it'd be cleaner to just give `worker_0` parts `[0, ..., 5]` and then `worker_1` parts `[6, ..., 11]`. I think we just divide `n_largest_eval_parts` by `n_parts` (I think?) when `n_largest_eval_parts > n_parts`, otherwise it can just be `parts_idx = j`. So `parts_idx=0` gets the first `k_i`-many parts, `parts_idx=1` gets the second `k_i`-many parts, etc. This actually already is the case when `(X_i, y_i)` is just the training set. I need to think a little more on this, but I don't think that would be that hard. What do you think, a better approach to divvying eval_set component parts?

Reply:
> Until now I guess I have been disregarding data shuffling almost entirely since there's no guarantee in any point of this dask module that certain workers will be dispatched any individual part of the training set

`lightgbm.dask` does prevent shuffling of the data, I think, by explicitly checking which workers hold different parts: https://github.com/microsoft/LightGBM/blob/72d7010b5ae09818541a749bd02b455b5a08de5b/python-package/lightgbm/dask.py#L350-L355.

But now that you mention it I'm actually not sure what the behavior would be if, for example, `data` and `label` had different partitioning. Maybe this would fail or maybe something bad would silently happen :grimacing: 

https://github.com/microsoft/LightGBM/blob/72d7010b5ae09818541a749bd02b455b5a08de5b/python-package/lightgbm/dask.py#L323

Either way, thanks for confirming that I have the right mental model here! **I don't think you should mess with this right now and that it's a good candidate for followup PRs.**

Your proposal makes sense to me as another way to chunk things up, but I don't think it changes the risk of shuffling. I can't think of a reason that `X_e.partitions[0]` and `data.partitions[0]` (for example) are more likely to be on the same worker than they are to be on different workers. Similarly, I can't think of a reason to believe that `X_e.partitions[0]` and `X_e.partitions[1]` would be more likely to be on the same worker than they are to be on different workers.